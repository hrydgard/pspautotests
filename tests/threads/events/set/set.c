#include "../sub_shared.h"

SETUP_SCHED_TEST;

#define SET_TEST(title, flag, bits) { \
	int result = sceKernelSetEventFlag(flag, bits); \
	if (result == 0) { \
		printf("%s: OK\n", title); \
	} else { \
		printf("%s: Failed (%X)\n", title, result); \
	} \
	PRINT_FLAG(flag); \
}

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedulingLogPos += vsprintf(schedulingLog + schedulingLogPos, format, args);
	va_end(args);
}

static int scheduleTestFunc1(SceSize argSize, void* argPointer) {
	u32 bits = 0xDEADBEEF;
	SceUInt timeout = 5000;

	schedf("T1B");
	int result = sceKernelWaitEventFlagCB(*(int*) argPointer, 0x00FF, PSP_EVENT_WAITOR | PSP_EVENT_WAITCLEAR, &bits, &timeout);
	schedf("T1D (result=%08x, bits=%08x) ", result, (uint) bits);

	return 0;
}

static int scheduleTestFunc2(SceSize argSize, void* argPointer) {
	int result = 0x800201A8;
	u32 bits = 0xDEADBEEF;
	SceUInt timeout;

	schedf("T2B");
	while (result == 0x800201A8) {
		timeout = 1;
		result = sceKernelWaitEventFlagCB(*(int*) argPointer, 0xFF00, PSP_EVENT_WAITOR | PSP_EVENT_WAITCLEAR, &bits, &timeout);
	}
	schedf("T2D (result=%08x, bits=%08x) ", result, (uint) bits);

	return 0;
}

static int scheduleTestFunc3(SceSize argSize, void* argPointer) {
	int result = 0x800201A8;
	u32 bits = 0xDEADBEEF;
	SceUInt timeout;

	schedf("T3B");
	while (result == 0x800201A8) {
		timeout = 1;
		result = sceKernelWaitEventFlagCB(*(int*) argPointer, 0xFFFF, PSP_EVENT_WAITOR | PSP_EVENT_WAITCLEAR, &bits, &timeout);
	}
	schedf("T3D (result=%08x, bits=%08x) ", result, (uint) bits);

	return 0;
}

int main(int argc, char **argv) {
	SceUID flag = sceKernelCreateEventFlag("set", 0, 0, NULL);
	PRINT_FLAG(flag);

	SET_TEST("Basic 0x2", flag, 0x2);
	SET_TEST("Basic 0x1", flag, 0x1);
	SET_TEST("Basic 0x1", flag, 0x1);
	SET_TEST("Basic 0xF000", flag, 0xF000);
	SET_TEST("Zero", flag, 0);

	sceKernelDeleteEventFlag(flag);

	flag = sceKernelCreateEventFlag("signal", 0, 0xFFFFFFFE, NULL);
	PRINT_FLAG(flag);
	SET_TEST("All but 0x1 + 0x1", flag, 0x1);
	sceKernelDeleteEventFlag(flag);

	SET_TEST("NULL", 0, 1);
	SET_TEST("Invalid", 0xDEADBEEF, 1);
	SET_TEST("Deleted", flag, 1);

	BASIC_SCHED_TEST("NULL",
		result = sceKernelSetEventFlag(0, 0);
	);
	BASIC_SCHED_TEST("Other + 2",
		result = sceKernelSetEventFlag(flag2, 2);
	);
	BASIC_SCHED_TEST("Other + 1",
		result = sceKernelSetEventFlag(flag2, 1);
	);
	BASIC_SCHED_TEST("Same + 2",
		result = sceKernelSetEventFlag(flag1, 2);
	);
	BASIC_SCHED_TEST("Same + 1",
		result = sceKernelSetEventFlag(flag1, 1);
	);

	BASIC_SCHED_TEST("Same + 1",
		result = sceKernelSetEventFlag(flag1, 1);
	);

	flag = sceKernelCreateEventFlag("set", 0x200, 0, NULL);
	schedulingLogPos = 0;
	schedf("A");
	SceUID thread1 = CREATE_SIMPLE_THREAD(scheduleTestFunc1);
	SceUID thread2 = CREATE_SIMPLE_THREAD(scheduleTestFunc2);
	SceUID thread3 = CREATE_SIMPLE_THREAD(scheduleTestFunc3);
	sceKernelStartThread(thread1, sizeof(flag), &flag);
	sceKernelStartThread(thread2, sizeof(flag), &flag);
	sceKernelStartThread(thread3, sizeof(flag), &flag);
	schedf("C");
	int result = sceKernelSetEventFlag(flag, 0xFFFF);
	schedf("E");
	sceKernelDeleteEventFlag(flag);
	schedf("F");

	printf("Set with multi and clear: %s(main=%08X)\n", schedulingLog, result);

	return 0;
}