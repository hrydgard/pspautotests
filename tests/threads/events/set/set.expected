Event flag: OK (size=52,name='set',attr=0,init=00000000,cur=00000000,wait=0)
Basic 0x2: OK
Event flag: OK (size=52,name='set',attr=0,init=00000000,cur=00000002,wait=0)
Basic 0x1: OK
Event flag: OK (size=52,name='set',attr=0,init=00000000,cur=00000003,wait=0)
Basic 0x1: OK
Event flag: OK (size=52,name='set',attr=0,init=00000000,cur=00000003,wait=0)
Basic 0xF000: OK
Event flag: OK (size=52,name='set',attr=0,init=00000000,cur=0000F003,wait=0)
Zero: OK
Event flag: OK (size=52,name='set',attr=0,init=00000000,cur=0000F003,wait=0)
Event flag: OK (size=52,name='signal',attr=0,init=FFFFFFFE,cur=FFFFFFFE,wait=0)
All but 0x1 + 0x1: OK
Event flag: OK (size=52,name='signal',attr=0,init=FFFFFFFE,cur=FFFFFFFF,wait=0)
NULL: Failed (8002019A)
Event flag: Failed (00000000)
Invalid: Failed (8002019A)
Event flag: Invalid (8002019A)
Deleted: Failed (8002019A)
Event flag: Invalid (8002019A)
NULL: A1B1C2E1D1F2 (thread=800201B5, thread bits=00000000, main=8002019A)
Other + 2: A1B1C2E1D1F2 (thread=800201B5, thread bits=00000000, main=00000000)
Other + 1: A1B1C2E1D1F2 (thread=800201B5, thread bits=00000000, main=00000000)
Same + 2: A1B1C2D1E2F1 (thread=00000000, thread bits=00000002, main=00000000)
Same + 1: A1B1C2D1E2F1 (thread=00000000, thread bits=00000001, main=00000000)
Same + 1: A1B1C2D1E2F1 (thread=00000000, thread bits=00000001, main=00000000)
Set with multi and clear: AT1BT2BT3BCT1D (result=00000000, bits=0000ffff) T2D (result=00000000, bits=0000ff00) ET3D (result=800201b5, bits=00000000) F(main=00000000)
